sklearn/ensemble/tests/test_forest.py

```diff
@@ -1330,3 +1330,93 @@ def test_forest_degenerate_feature_importances():
     gbr = RandomForestRegressor(n_estimators=10).fit(X, y)
     assert_array_equal(gbr.feature_importances_,
                        np.zeros(10, dtype=np.float64))
+
+
+def test__get_n_bootstrap_samples():
+    """ Test bootstrap sample number getter
+    """
+
+    from sklearn.ensemble.forest import _get_n_bootstrap_samples
+
+    n_samples = 10
+
+    # Check that max_samples=None defaults to full n_samples
+    n_bootstrap = _get_n_bootstrap_samples(n_samples, max_samples=None)
+    assert n_bootstrap == n_samples
+
+    # Check that max less total yields max
+    n_bootstrap = _get_n_bootstrap_samples(n_samples, max_samples=5)
+    assert n_bootstrap == 5
+
+    # Check that indices with max_samples kwarg subsamples by that amount
+    n_bootstrap = _get_n_bootstrap_samples(n_samples, max_samples=0.1)
+    assert n_bootstrap == 1
+
+    # Check that ValueError is raised when `max_samples` is integral
+    # and greater than `n_samples`
+    with pytest.raises(ValueError):
+        _get_n_bootstrap_samples(n_samples, max_samples=n_samples + 1)
+
+    # Check that ValueError is raised when `max_samples` is float
+    # and not in range (0, 1)
+    with pytest.raises(ValueError):
+        # Edge case at 1.0
+        _get_n_bootstrap_samples(n_samples, max_samples=1.0)
+    with pytest.raises(ValueError):
+        # Greater than 1.0
+        _get_n_bootstrap_samples(n_samples, max_samples=2.0)
+    with pytest.raises(ValueError):
+        # Edge case at 0.0
+        _get_n_bootstrap_samples(n_samples, max_samples=0.0)
+    with pytest.raises(ValueError):
+        # Less than 0.0
+        _get_n_bootstrap_samples(n_samples, max_samples=-1.0)
+    with pytest.raises(ValueError):
+        # NaN: valid float, so should raise ValueError
+        _get_n_bootstrap_samples(n_samples, max_samples=np.nan)
+    with pytest.raises(ValueError):
+        # Inf: valid float, so should raise ValueError
+        _get_n_bootstrap_samples(n_samples, max_samples=np.inf)
+
+    # Check that TypeError is raised when `max_samples` is the wrong type
+    with pytest.raises(TypeError):
+        _get_n_bootstrap_samples(n_samples, max_samples='bad max sample type')
+    with pytest.raises(TypeError):
+        _get_n_bootstrap_samples(n_samples, max_samples=np.ones(n_samples))
+
+
+def check_classification_toy_max_samples(name):
```

Comment:
A more obvious test to me would be to fit a couple of trees on a big dataset with a small number of samples (`max_samples=0.1`) and on the full dataset (`max_samples=None`). I would expect to have much more nodes in the second case than the first case.

Reply:
That was my initial inclination as well, but I wasn't sure how to make "much more nodes" quantitative. This is related to https://github.com/scikit-learn/scikit-learn/pull/14682#discussion_r315960384.

Reply:
"much more nodes" could only be "bigger than" (>). This should be enough.

Reply:
I just noticed that the 32bit CI is failing on one of these tests, but passing on the rest. On further inspection, I notice on my local (64 bit) machine:
```
                                   Node Count
Method                  max_samples=None  max_samples=2
-------------------------------------------------------
RandomForestRegressor   4643              1
ExtraTreesRegressor     13165             13121
RandomForestClassifier  5377              1
ExtraTreesClassifier    15503             15373
```
In other words, the unit test is performing as expected for `Random*` and passing by sheer coincidence of my chosen random seed for `Extra*`. And I assume because of slight variations in floating point on the 32 bit machine, this leads to failure in the CI.

It seems like this type of test should not be run for the `Extra*` family, but I'm not familiar enough with the method to explain why. Thoughts, @glemaitre?

Reply:
Yep, we can limit the test to RandomTree then.

Reply:
https://github.com/scikit-learn/scikit-learn/pull/14682/commits/9d72d36c2fe9ebb96892308b17cd1ae72809c555